### PR TITLE
Allow limiting the length of the sql string that will be printed

### PR DESCRIPTION
**Pull request by thiagotalma**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes

When we run a migration with a very large view, for example, it is bad to follow the migration progress on the screen.

### CODE REVIEW

**Comment by samdark** [comment]
Could you elaborate about use case?

**Comment by thiagotalma** [comment]
I have a database view that takes up 4 pages of the screen when it is executed in a migration.
It would be very useful to be able to configure to only print the first 200 characters.

**Comment by samdark** [comment]
I suggest making it a global class property instead of a method parameter.

**Review by samdark — APPROVED** [review]

File: framework/db/Migration.php [hunk 1]
```diff
@@ -184,7 +189,8 @@ public function safeDown()
      */
     public function execute($sql, $params = [])
     {
-        echo "    > execute SQL: $sql ...";
+        $sqlPrint = ($this->printSqlLength === 0 || strlen($sql) <= $this->printSqlLength) ? $sql : substr($sql, 0, $this->printSqlLength) . "[...hidden]";
```
Comment (dynasource) [inline framework/db/Migration.php]: difficult line to read. Deserves a multiple lines.

**Review by dynasource — APPROVED** [review]

File: framework/db/Migration.php (hunk 1 above)
Comment (rob006) [inline framework/db/Migration.php]: You probably should use `mb_strlen()` and `mb_substr()` here.

**Review by samdark — CHANGES_REQUESTED** [review]
See comments.

File: framework/db/Migration.php [hunk 2]
```diff
@@ -184,7 +189,13 @@ public function safeDown()
      */
     public function execute($sql, $params = [])
     {
-        echo "    > execute SQL: $sql ...";
+        $sqlPrint = $sql;
+
+        if ($this->printSqlLength !== 0 && mb_strlen($sql) > $this->printSqlLength) {
```
Comment (samdark) [inline framework/db/Migration.php]: `mb_strlen` should have encoding specified.

File: framework/db/Migration.php [hunk 3]
```diff
@@ -184,7 +189,13 @@ public function safeDown()
...
+            $sqlPrint = mb_substr($sql, 0, $this->printSqlLength) . "[...hidden]";
```
Comment (samdark) [inline framework/db/Migration.php]: Same here. Encoding should be specified.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/db/Migration.php (hunk 2 above)
Comment (thiagotalma) [inline framework/db/Migration.php]: Which one should I use?
I searched for a pattern in the framework and did not find it.
In fact, most cases are without encoding.

Comment (samdark) [inline framework/db/Migration.php]: mb_strlen($from, Yii::$app->charset)

Comment (rob006) [inline framework/db/Migration.php]: `StringHelper::truncate()` may be better option in this case.

Comment (samdark) [inline framework/db/Migration.php]: Actually, yes.

File: framework/db/Migration.php [hunk 4]
```diff
@@ -61,6 +61,11 @@ class Migration extends Component implements MigrationInterface
      */
     public $db = 'db';
 
+    /**
+     * @var int length of the string that will be printed on the screen.
+     */
+    public $printSqlLength = 0;
```
Comment (bashkarev) [inline framework/db/Migration.php]: - `-1` - show all
- `0` - not display 
?

Comment (dynasource) [inline framework/db/Migration.php]: agree that 0 is counter intuitive

Comment (rob006) [inline framework/db/Migration.php]:
```php
public $printSqlLengthLimit;
```

If `null`, there is no limit. If `0`, query id not displayed.

File: framework/db/Migration.php [hunk 5]
```diff
@@ -61,6 +62,10 @@ class Migration extends Component implements MigrationInterface
      */
     public $db = 'db';
 
+    /**
+     * @var int length of the string that will be printed on the screen.
+     */
+    public $printSqlLength;
```
Comment (arogachev) [inline framework/db/Migration.php]: I suggest using `$printedSqlLength` or `$displayedSqlLength` instead.

File: framework/db/Migration.php [hunk 6]
```diff
@@ -184,7 +189,13 @@ public function safeDown()
      */
     public function execute($sql, $params = [])
     {
-        echo "    > execute SQL: $sql ...";
+        $sqlPrint = $sql;
```
Comment (arogachev) [inline framework/db/Migration.php]: `$printedSql` is better, more clear and correct form IMO.

File: framework/db/Migration.php [hunk 7]
```diff
@@ -184,7 +189,13 @@ public function safeDown()
...
+
+        if ($this->printSqlLength !== null) {
+            $sqlPrint = StringHelper::truncate($sql, $this->printSqlLength, '[...hidden]');
```
Comment (arogachev) [inline framework/db/Migration.php]: Nitpicking, but maybe add a space after dots: `[... hidden]`.

**Review by arogachev — CHANGES_REQUESTED** [review]

**Comment by thiagotalma** [comment]
@arogachev done!

File: framework/db/Migration.php (hunk 3 above)
Comment (thiagotalma) [inline framework/db/Migration.php]: @samdark done!

**Review by samdark — APPROVED** [review]

**Comment by arogachev** [comment]
@thiagotalma Thanks! 👍 Made small corrections. Now I think it's ready for merge. Others,, please check.

**Review by arogachev — APPROVED** [review]

**Comment by samdark** [comment]
Merged with adjustments. Thanks!

**Pull request closed without merging**